opm/simulators/wells/MultisegmentWellGeneric.cpp

```diff
@@ -622,6 +530,185 @@ computeBhpAtThpLimitProd(const std::function<std::vector<double>(const double)>&
+                high = curr;
+                f_high = f_curr;
+            }
+            ++it;
+        }
+        bhp_max = low;
+    }
+    deferred_logger.debug("computeBhpAtThpLimitProd(): well = " + baseif_.name() +
+                          "  low = " + std::to_string(low) +
+                          "  high = " + std::to_string(high) +
+                          "  f(low) = " + std::to_string(f_low) +
+                          "  f(high) = " + std::to_string(f_high) +
+                          "  bhp_max = " + std::to_string(bhp_max));
+    return bhp_max;
+}
+
+
+template<typename Scalar>
+bool
+MultisegmentWellGeneric<Scalar>::
+bisectBracket(const std::function<double(const double)>& eq,
+              const std::array<double, 2>& range,
+              double& low, double& high,
+              std::optional<double>& approximate_solution,
+              DeferredLogger& deferred_logger) const
+{
+    bool finding_bracket = false;
+    low = range[0];
+    high = range[1];
+
+    double eq_high = eq(high);
+    double eq_low = eq(low);
+    const double eq_bhplimit = eq_low;
+    deferred_logger.debug("computeBhpAtThpLimitProd(): well = " + baseif_.name() +
+                          "  low = " + std::to_string(low) +
+                          "  high = " + std::to_string(high) +
+                          "  eq(low) = " + std::to_string(eq_low) +
+                          "  eq(high) = " + std::to_string(eq_high));
+    if (eq_low * eq_high > 0.0) {
+        // Failed to bracket the zero.
+        // If this is due to having two solutions, bisect until bracketed.
+        double abs_low = std::fabs(eq_low);
+        double abs_high = std::fabs(eq_high);
+        int bracket_attempts = 0;
+        const int max_bracket_attempts = 20;
+        double interval = high - low;
+        const double min_interval = 1.0 * unit::barsa;
+        while (eq_low * eq_high > 0.0 && bracket_attempts < max_bracket_attempts && interval > min_interval) {
+            if (abs_high < abs_low) {
+                low = 0.5 * (low + high);
+                eq_low = eq(low);
+                abs_low = std::fabs(eq_low);
+            } else {
+                high = 0.5 * (low + high);
+                eq_high = eq(high);
+                abs_high = std::fabs(eq_high);
+            }
+            ++bracket_attempts;
+        }
+
+        if (eq_low * eq_high <= 0.) {
+            // We have a bracket!
+            finding_bracket = true;
+            // Now, see if (bhplimit, low) is a bracket in addition to (low, high).
+            // If so, that is the bracket we shall use, choosing the solution with the
+            // highest flow.
+            if (eq_low * eq_bhplimit <= 0.0) {
+                high = low;
+                low = range[0];
+            }
+        } else { // eq_low * eq_high > 0.0
+            // Still failed bracketing!
+            const double limit = 3.0 * unit::barsa;
+            if (std::min(abs_low, abs_high) < limit) {
```

Comment:
A minor dilemma here: instead of returning the least bad solution, we could decide to continue with the brute force bracketing.

Reply:
Yes. The reason for this is that I did not want to introduce extra changes in the logic for easy testing.  I will handle it in a separate PR by removing this usage and test it there.

So far, it looks like the brute force is relatively reliable in term of finding the solution, which is a finding due to the bug introduced by the refactoring process.

Reply:
Except this comment, other comments are addressed.  I will create a new PR for this comment in a separate PR after this PR goes in for easy testing.